From: Chris Mason <chris.mason@fusionio.com>
Date: Wed, 25 Jul 2012 16:03:32 -0400
Patch-mainline: 3.6
Git-commit: cd1cfc49153ba2bef247e500d8bd4d135193ece9
References: FATE#312888
Subject: [PATCH] Btrfs: add a barrier before a waitqueue_active check

We were missing wakeups on the delayed ref waitqueue due
to races on waitqueue_active.

Signed-off-by: Chris Mason <chris.mason@fusionio.com>
Signed-off-by: David Sterba <dsterba@suse.cz>
---
 fs/btrfs/extent-tree.c |    1 +
 1 file changed, 1 insertion(+)

--- a/fs/btrfs/extent-tree.c
+++ b/fs/btrfs/extent-tree.c
@@ -5298,6 +5298,7 @@ static noinline int check_ref_cleanup(st
 	rb_erase(&head->node.rb_node, &delayed_refs->root);
 
 	delayed_refs->num_entries--;
+	smp_mb();
 	if (waitqueue_active(&root->fs_info->tree_mod_seq_wait))
 		wake_up(&root->fs_info->tree_mod_seq_wait);
 
